kernel/src/sched.rs

```diff
@@ -284,63 +309,33 @@ impl Kernel {
         }
     }
 
-    /// Main loop.
-    pub fn kernel_loop<P: Platform, C: Chip>(
-        &'static self,
-        platform: &P,
-        chip: &C,
-        ipc: Option<&ipc::IPC>,
-        _capability: &dyn capabilities::MainLoopCapability,
-    ) {
-        loop {
-            unsafe {
-                chip.service_pending_interrupts();
-                DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
-
-                for p in self.processes.iter() {
-                    p.map(|process| {
-                        self.do_process(platform, chip, process, ipc);
-                    });
-                    if chip.has_pending_interrupts()
-                        || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                    {
-                        break;
-                    }
-                }
-
-                chip.atomic(|| {
-                    if !chip.has_pending_interrupts()
-                        && !DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                        && self.processes_blocked()
-                    {
-                        chip.sleep();
-                    }
-                });
-            };
-        }
-    }
-
-    unsafe fn do_process<P: Platform, C: Chip>(
+    unsafe fn do_process<P: Platform, C: Chip, S: SysTick>(
         &self,
         platform: &P,
         chip: &C,
+        systick: &S,
         process: &dyn process::ProcessType,
         ipc: Option<&crate::ipc::IPC>,
-    ) {
-        let systick = chip.systick();
+        timeslice_us: u32,
+        break_for_kernel_tasks: bool,
+    ) -> (StoppedExecutingReason, u32) {
         systick.reset();
-        systick.set_timer(KERNEL_TICK_DURATION_US);
+        systick.set_timer(timeslice_us);
```

Comment:
Comment really for the next line.

I know this is "how it was", but I wonder if we need to enforce some minimum `timeslice_us` value. My concern is the universe where this value is small, and the timeslice manages to expire before the switch to userland completes. As I see it, we have two opportunities to miss the systick interrupt:
 (1) It expires before we call `enable(true)` below, so the timer expires without triggering the interrupt
 (2) It expires *during* the switch to user (i.e. after the `enable(true)` call but before the context switch actually occurs); the interrupt handler will mark the systick expired, but then we're back to the context switch code effectively without a systick set.

Reply:
Thinking about the later `systick.enable(true)` more, I think we need to clarify the systick HIL a bit:

```
    /// Enables the timer
    ///
    /// Enabling the timer will begin a count down from the value set with
    /// `set_timer`.
    ///
    ///   * `with_interrupt` - if set, an expiring timer will fire an interrupt.
    fn enable(&self, with_interrupt: bool);
```

It's unclear to me from that language if the second call to `enable` will *reset* the systick timer (I think it will?) or simply enable interrupts.

Since the process can stay in the loop here depending on the syscall, it could effectively grant itself infinite time.

I think my instinct is that there should be exactly one call to `systick.enable` at the beginning of `do_process`

Reply:
> Comment really for the next line.
> 
> I know this is "how it was", but I wonder if we need to enforce some minimum `timeslice_us` value. My concern is the universe where this value is small, and the timeslice manages to expire before the switch to userland completes. As I see it, we have two opportunities to miss the systick interrupt:
> (1) It expires before we call `enable(true)` below, so the timer expires without triggering the interrupt
> (2) It expires _during_ the switch to user (i.e. after the `enable(true)` call but before the context switch actually occurs); the interrupt handler will mark the systick expired, but then we're back to the context switch code effectively without a systick set.

I think that the check on `systick.greater_than(MIN_QUANTA_THRESHOLD_US)` effectively enforces a minimum timeslice of `MIN_QUANTA_THRESHOLD_US`

Reply:
> Thinking about the later `systick.enable(true)` more, I think we need to clarify the systick HIL a bit:
> 
> ```
>     /// Enables the timer
>     ///
>     /// Enabling the timer will begin a count down from the value set with
>     /// `set_timer`.
>     ///
>     ///   * `with_interrupt` - if set, an expiring timer will fire an interrupt.
>     fn enable(&self, with_interrupt: bool);
> ```
> 
> It's unclear to me from that language if the second call to `enable` will _reset_ the systick timer (I think it will?) or simply enable interrupts.
> 
> Since the process can stay in the loop here depending on the syscall, it could effectively grant itself infinite time.
> 
> I think my instinct is that there should be exactly one call to `systick.enable` at the beginning of `do_process`

The cortex-m implementation does not reset the timer, from looking at the register writes. I think that is the correct behavior, I will update the documentation to reflect that.

I do not think we want the possibility of the systick firing while we are in the kernel -- I do not totally understand the systick_handler assembly, but the documentation seems to indicate it redirects control flow to the switch to user function, and I do not think we want that to be possible while we are in the kernel. I think calling enable twice seems to make sense, though perhaps the systick trait could be changed to have seperate start_counter() and enable_interrupts() functions.

Reply:
> I do not think we want the possibility of the systick firing while we are in the kernel

This possibility does exist. I was doing something else with our kernel, when I discovered it few days back. [Here](https://github.com/tock/tock/blob/313a5e12298b83b49cb95dd5d39baa5332bb1a13/arch/cortex-m/src/syscall.rs#L164) we might come out of the kernel because of an interrupt or a syscall, while the systick timer is still running. So, we'll capture `systick_expired` to be `0`. Then by the time we get to stopping systick timer, the systick interrupt could have fired, and `Systick_CurrentValue:CURRENT[23:0]` would have `Systick_Reload:RELOAD[23:0]`. So, the remaining time slice that gets captured *could* be wrong. Capturing remaining time-slice accurately is important for our use-case.

While I've not coded the implementation yet, to me it seems like the correct approach should be that we stop the systick timer in assembly within `svc_handler`, `generic_isr` and `systick_handler`, so we can accurately capture `Systick_CurrentValue:CURRENT[23:0]` value.

Reply:
Oy, I had not yet even considered the race between systick and other interrupt handlers. Here lies very subtle and rare bugs :(.

On ARM at least, I think it should be viable to configure the systick to be lower priority than other interrupts, so we can be sure that a systick interrupt won't fire while other handlers are running. I don't know enough about Risc-V to say for certain whether the same fix is viable there.

More generally, I'm concerned about how we express this pretty low-level requirement in a robust, cross-platform way such that all chips / arches get it right.

Reply:
> > I do not think we want the possibility of the systick firing while we are in the kernel
> 
> 
> 
> This possibility does exist. I was doing something else with our kernel, when I discovered it few days back. [Here](https://github.com/tock/tock/blob/313a5e12298b83b49cb95dd5d39baa5332bb1a13/arch/cortex-m/src/syscall.rs#L164) we might come out of the user-space because of an interrupt or a syscall, while the systick timer is still running. So, we'll capture `systick_expired` to be `0`. Then by the time we get to stopping systick timer, the systick interrupt could have fired, and `Systick_CurrentValue:CURRENT[23:0]` would have `Systick_Reload:RELOAD[23:0]`. So, the remaining time slice that gets captured *could* be wrong. Capturing remaining time-slice accurately is important for our use-case.
> 
> 
> 
> While I've not coded the implementation yet, to me it seems like the correct approach should be that we stop the systick timer in assembly within `svc_handler`, `generic_isr` and `systick_handler`, so we can accurately capture `Systick_CurrentValue:CURRENT[23:0]` value.
> 
> 

Hey @rajivr , in order to accurately capture the amount of time used, it sounds like you must already be using a modified scheduler? If so, is this interface sufficient to capture the behavior you want from your scheduler? 

I agree that currently it is possible for the reload value to be read by systick.get_value() in the case of the race you describe —fortunately the cortex-m implementation sets the reload value to be 0 until set_value() is called, so I think it captures the equivalent of if the systick interrupt had occurred before the return from userspace. Does this sound right?

That being said, my opinion is that it makes sense to charge syscall handling that occurs in the kernel to the process. If we stop the systick in svc_handler, I am not sure we can do so. Do you have a different opinion on this? Also there is the problem that I believe the risc-v systick does not support a mechanism for pausing the timer.

Reply:
> Hey @rajivr , in order to accurately capture the amount of time used, it sounds like you must already be using a modified scheduler? If so, is this interface sufficient to capture the behavior you want from your scheduler?

Sorry for the delay in replying.

Our scheduler is still under development. We are taking a different approach with our scheduler. Our scheduler interface is largely inspired from,

https://github.com/xen-project/xen/blob/stable-4.12/xen/include/xen/sched-if.h#L159-L169

You can replace `vcpu` with `process`, and it provides a rather minimal scheduler interface to the rest of the kernel.

`do_schedule` is where most of the scheduler policy decisions can be taken. In our use-case, we'll only be implementing one scheduler, but Xen has multiple scheduler implementations. In case you are curious, check out the files `sched_{credit, credit2, null, rt}.c` in the tree.

> That being said, my opinion is that it makes sense to charge syscall handling that occurs in the kernel to the process. If we stop the systick in svc_handler, I am not sure we can do so. Do you have a different opinion on this? Also there is the problem that I believe the risc-v systick does not support a mechanism for pausing the timer.

While RISC-V is not our immediate roadmap, I did check out the documentation for `mtime` and `mtimecmp`, when I was thinking about the design. I do not have hands on experience with RISC-V, so please the next comment with a grain of salt. 

I don't think we necessarily have to stop the timer in the trap handler. In case of RISC-V the approach that could potentially work would be to disable the `MTIE` bit in the trap handler, and then copy out the `mtime` value into a known memory location.

So, if we returned back to the kernel due to a system call or an interrupt, we can then check this known memory location and compute how much time was spent in the user-mode.

Reply:
One issue is that 2 of our 3 currently supported RISC-V chips only have a single hardware timer (the mtimer), so for these chips we will likely need to implement the systick as an alarm on top of an AlarmMux that shares a hardware timer with other alarms in the kernel. This makes disabling MTIE while in the kernel unpalatable, because it is also required for interrupts unrelated to timeslice expirations

Reply:
> One issue is that 2 of our 3 currently supported RISC-V chips only have a single hardware timer (the mtimer), so for these chips we will likely need to implement the systick as an alarm on top of an AlarmMux that shares a hardware timer with other alarms in the kernel. This makes disabling MTIE while in the kernel unpalatable, because it is also required for interrupts unrelated to timeslice expirations

OpenTitan [Earlgrey](https://docs.opentitan.org/hw/top_earlgrey/doc/) does seem to have only one timer. I think having just one timer peripheral is very limiting. I would still prefer having a dedicated timer for the scheduler. Virtual timer would add additional layer of complexity into an important code-path in the kernel.

Reply:
> OpenTitan [Earlgrey](https://docs.opentitan.org/hw/top_earlgrey/doc/) does seem to have only one timer. I think having just one timer peripheral is very limiting. I would still prefer having a dedicated timer for the scheduler. Virtual timer would add additional layer of complexity into an important code-path in the kernel.

The sifive e21 has the same limitation. I am hoping I can come up with an abstraction where most boards use a dedicated timer, but boards without one can implement their own lightweight virtualization and pass a SysTick to the Kernel that behaves like a dedicated timer.

Reply:
Marking this conversation as resolved. `MIN_QUANTA_THRESHOLD_US` presents as an effective minimum timeslice, and the possible race condition for detecting timeslice usage should be resolved by #1985 .